### ISSUE TITLE
UseNewSymlinkResolver causes applications to create both the link and the target folder

### ISSUE THREAD

**Issue by offhub** [sandboxie-plus/Sandboxie#3481]
### Describe what you noticed and did

UseNewSymlinkResolver causes applications to create both the link and the target folder in the sandbox. As a result, changes made to applications may not be saved properly.

1. Create a symlink for Edge\User Data folder

from unsandboxed cmd:
```
cd /d "%LOCALAPPDATA%\Microsoft\Edge"
xcopy /e /y "User Data\" "%TEMP%\User Data\"
ren "User Data" "User Data.BAK"
mklink /D "%LOCALAPPDATA%\Microsoft\Edge\User Data" "%TEMP%\User Data"
```
2. Run Ms Edge in sandbox
3. The "User Data" folder is created in both locations within the sandbox.
4. Use Windows Explorer to browse to both locations.

from sandboxed Windows Explorer:
```
%TEMP%
%LOCALAPPDATA%\Microsoft\Edge
```

from unsandboxed Windows Explorer:

```
C:\Sandbox\user\DefaultBox\drive\C\Users\user\AppData\Local\Temp\User Data
C:\Sandbox\user\DefaultBox\user\current\AppData\Local\Microsoft\Edge\User Data
```

### How often did you encounter it so far?

Every time

### Affected program

msedge, firefox, etc.

### Download link

Not relevant

### Where is the program located?

The program is installed only outside the sandbox.

### Expected behavior

A symbolic link folder should only be created in a single location within the sandbox.

### What is your Windows edition and version?

Windows 10 Pro 22H2 64-bit (19045.3693)

### In which Windows account you have this problem?

A local account (Administrator)., A Microsoft account (Administrator)., An account with UAC protection set to Always notify.

### Please mention any installed security software

Microsoft Windows Defender

### What version of Sandboxie are you running?

Sandboxie-Plus 1.12.3 64-bit

### Is it a new installation of Sandboxie?

I recently did a new clean installation.

### In which sandbox type you have this problem?

All sandbox types (I tried them all).

### Can you reproduce this problem on a new empty sandbox?

I can confirm it also on a new empty sandbox.

### Sandboxie.ini configuration

```shell
UseNewSymlinkResolver=y
```

**Comment by profucius** [comment]
I believe I've found a regression on this ticket. 

Situation: I use a specific app (Obsidian, a markdown editor) within Sandboxie to create "profiles" of that app, so that I can run multiple instances of it at once. Instead of duplicating the data between the profiles, I use symbolic links to share folder access between all of them.

On Sandboxie 1.12.3, this worked perfectly fine. 
However, on 1.12.4, this functionality breaks.

On .4, the app Obsidian (ran within the sandbox), it seems that Obsidian is seeing the symbolic linked folder as a folder, but it does not see any content within it. On .3, all content within showed fine.

Nothing else changed within Obsidian or the symbolic links. The only change made in my testing was installing the update for .3 to .4

Below I have provided a couple of screen captures demonstrating this issue. I tried to make them appear as exact in every other way. Take note of the Sandboxie version, and within the Obsidian app left sidebar, the "Symlink" (folder) toggles open and shows a file within .3 but within .4 it toggles open empty.

Let me know if I can provide any other information. If I need to open a new ticket for this, I will do that too. Thank you.

# Working on 1.12.3
https://github.com/sandboxie-plus/Sandboxie/assets/18254607/0316e1b5-965d-4e50-9982-efe0be00d6b6

# Broken on 1.12.4
https://github.com/sandboxie-plus/Sandboxie/assets/18254607/1ea7121e-853b-4db8-954e-0ea3e20670a1

**Comment by isaak654** [comment]
For the record, I cannot play both videos on Microsoft Edge and Firefox, but fortunately it is possible to save them locally.

**Comment by DavidXanatos** [comment]
what can you play them with my VLC fails

**Comment by DavidXanatos** [comment]
@profucius could you write down a manual test case how to create the sym links and how to test without the software,
perhaps using cmd.exe only

**Comment by isaak654** [comment]
> what can you play them with my VLC fails

Both videos work fine with K-Lite Codec Pack Mega.

**Comment by profucius** [comment]
Apologies for the videos not working for you. I converted them using Adobe Media Encoder to reduce the file size, using one of the built-in codecs. It uploaded fine and plays in my browser fine, so I didn't realize it would be an issue. Glad you could at least view them via download though. PS- If you can recommend a better screen capture and upload tool, I'll take a look at it.

As for replicating the issue, I used multiple tools to create the symbolic links. Windows has its own built-in cmd line commands, and there is a [shell extension](https://schinagl.priv.at/nt/hardlinkshellext/linkshellextension.html#contact) tool that I prefer. This makes it so you can copy a file/folder and then go to the destination, right click and tell it to create a symbolic link there. To be clear, I am only creating symbolic links, not junctions or hard links or any other type.

Windows Explorer loaded within the sandbox is able to see through the symbolic links I created as expected. However the Obsidian app demonstrated in my video is what is now unable to do so, as of 1.12.4.

I hope that is the information you're looking for, but if you need anything else let me know. It may only be possible to reproduce the issue by installing [Obsidian](https://obsidian.md/) within a sandbox for yourself. It is a free software.

**Comment by profucius** [comment]
Update: I just tried an idea I had and it gave me a rather bizarre result (to me). Perhaps this will help in diagnosis.

I created an empty folder and named "Symbolic Link". I placed that folder at the root of the same partition as the Sandboxie sandbox. Inside that empty folder I just created an empty file called File.txt. I duplicated this folder and placed it at the root of a different partition. So now there are two exact same dummy folders and files, but on different partitions. And I created a symbolic link of each folder into the sandbox, I named the one on the same partition "Symbolic Link Same" and the one on the different partition "Symbolic Link Different"

On Sandboxie 1.12.3, the symbolic link created from the folder on the different partition is showing data as expected, but the symbolic link created from the folder on the same partition is not showing anything at all, as if it were empty.

What is bizarre is that on 1.12.4, this is reversed. The different one shows nothing, and the same one shows correctly. Screenshots below to show what I mean.

Again I'm quite confused by this, but perhaps it will help in diagnosis. Let me know if I can provide anything else.

# 1.12.3
![1 12 3](https://github.com/sandboxie-plus/Sandboxie/assets/18254607/88ec0808-220f-4426-9c82-7faf81cc122f)

# 1.12.4
![1 12 4](https://github.com/sandboxie-plus/Sandboxie/assets/18254607/7ea0043a-db2f-4a0f-99d5-802bd8e9d478)

**Comment by DavidXanatos** [comment]
So you created the saylinks with an unsandboxed explorer, correct?

**Comment by profucius** [comment]
> So you created the saylinks with an unsandboxed explorer, correct?

Correct, the symlinks are created outside of the sandbox / on the host OS. The above findings are within the sandbox.

**Comment by DavidXanatos** [comment]
I created the links like you and cant reproduce the issue, 
could yo please post your sandboxie.ini

**Comment by profucius** [comment]
```
#
# Sandboxie configuration file
#

[GlobalSettings]

[UserSettings_0D16022D]
SbieCtrl_AutoStartAgent=SandMan.exe -autorun
BoxGrouping=:DefaultBox

[DefaultBox]
Enabled=y
BlockNetworkFiles=y
RecoverFolder=%{374DE290-123F-4565-9164-39C4925E467B}%
RecoverFolder=%Personal%
RecoverFolder=%Desktop%
BorderColor=#00FFFF,ttl
Template=OpenBluetooth
Template=SkipHook
Template=FileCopy
Template=qWave
Template=BlockPorts
Template=LingerPrograms
Template=AutoRecoverIgnore
ConfigLevel=10
```

**Comment by DavidXanatos** [comment]
well well that looks all right, strange that I cant reproduce it...
could you please test if the issue appears also with a completely empty sandbox?

**Comment by DavidXanatos** [comment]
https://github.com/sandboxie-plus/Sandboxie/assets/3890945/3cb5b28c-72e7-4961-9148-cdadb90d8208

for me it works, what am I doing wrong?

**Comment by DavidXanatos** [comment]
@profucius how do we continue?
I cant reproduce the issue see my video.

**Comment by profucius** [comment]
Apologies it's been a busy few days. I am surprised you weren't able to reproduce the issue. It looks like you did things correctly in the video. I tried it on two local machines of mine and both had the issue. I'll try again soon when I have a chance and post an update. If I can't find a way to detail a reproduction then I'll just assume at that point it's something on my end. Thanks

**Comment by profucius** [comment]
I re-watched your video and noticed you were installing Obsidian directly into the sandbox. I actually am not doing that, I install Obsidian on the host OS and run it within the sandbox. I tried reproducing the results by installing Obsidian directly into the sandbox like you did, but it did not make a difference. I wonder though, if you installed Obsidian within the host OS, would it change anything for you?

**Comment by isaak654** [comment]
Personally I'm able to reproduce this issue, so I can confirm the same conclusion:

> This demonstrates that the issue with Sandboxie is that it is passing data through symbolic links properly when the source folder is on the same partition as the sandbox, but it does not do this properly when the source folder is on a different partition as the sandbox.

I would like to point out the following differences from the [previous video](https://github.com/sandboxie-plus/Sandboxie/issues/3481#issuecomment-1904593792):

1) I have installed Obsidian in the real system (Windows 11 23H2 build 22631.3007) to be run in a new standard sandbox
2) I have dropped the symbolic links in `C:\Sandbox\User\Box\drive\C\MyVault`

**Comment by isaak654** [comment]
I have just reproduced this issue with Sandboxie Plus 1.12.7 x64, so I can confirm the same conclusion:

> This demonstrates that the issue with Sandboxie is that it is passing data through symbolic links properly when the source folder is on the same partition as the sandbox, but it does not do this properly when the source folder is on a different partition as the sandbox.

I would like to point out the following differences from the [previous video](https://github.com/sandboxie-plus/Sandboxie/issues/3481#issuecomment-1904593792):

1) I have installed Obsidian in the real system (Windows 11 23H2 x64 build 22631.3007) to be run in a new standard sandbox
2) I have dropped the symbolic links in `C:\Sandbox\User\Box\drive\C\MyVault`

**Comment by isaak654** [comment]
I have just reproduced this issue with Sandboxie Plus 1.12.7 x64, so I can confirm the same conclusion:

> This demonstrates that the issue with Sandboxie is that it is passing data through symbolic links properly when the source folder is on the same partition as the sandbox, but it does not do this properly when the source folder is on a different partition as the sandbox.

I would like to point out the following differences from the [previous video](https://github.com/sandboxie-plus/Sandboxie/issues/3481#issuecomment-1904593792):

1) I have installed Obsidian in the real system (Windows 11 23H2 x64 build 22631.3007) to be run in a new standard sandbox
2) I have dropped the symbolic links in `C:\Sandbox\User\Box\drive\C\MyVault`
3) I have created .md files, not .txt files

**Comment by isaak654** [comment]
I have just reproduced this issue with Sandboxie Plus 1.12.7 x64 and Windows 11 23H2 x64 build 22631.3007, so I can confirm the same conclusion:

> This demonstrates that the issue with Sandboxie is that it is passing data through symbolic links properly when the source folder is on the same partition as the sandbox, but it does not do this properly when the source folder is on a different partition as the sandbox.

I would like to point out the following differences from the [previous video](https://github.com/sandboxie-plus/Sandboxie/issues/3481#issuecomment-1904593792):

1) I have installed Obsidian outside the sandbox
2) I have dropped the symbolic links in `C:\Sandbox\User\Box\drive\C\MyVault`
3) I have created .md files, not .txt files

**Comment by DavidXanatos** [comment]
> I have dropped the symbolic links in C:\Sandbox\User\Box\drive\C\MyVault

That was it!!!!

When the symlinks exist outside the sandbox they work fine, when they live inside they do not.

When creating symlinks inside the sandbox they targets are supposed to point to the copy path not to the real path.
Injecting a symlink into a sandbox foodler structure breaks this rule.
That said, it whould be doable to make this work correctly even with symlinks pointing to true path locations.


Apparently I screwed up step 7

> Create a symbolic link of that folder and drop it into the Obsidian vault folder within the sandbox, should be the one you created at C:\

when testing

**Comment by DavidXanatos** [comment]
Could all please test the latest CI build to see if all works well now with symlinks

**Comment by profucius** [comment]
> Could all please test the latest CI build to see if all works well now with symlinks

How do I find this? I looked around the repo but I am not seeing it

**Comment by DavidXanatos** [comment]
https://github.com/sandboxie-plus/Sandboxie/actions/runs/7687015021

**Comment by profucius** [comment]
> https://github.com/sandboxie-plus/Sandboxie/actions/runs/7687015021

I downloaded the x64 package from that link. Uninstalled current sandboxie. Extracted zip and ran the SandMan.exe. It asked to do setup. Then gives me this error. I've tried manually starting the driver from the maintenance menu and it still tells me this error. Am I doing something wrong? This is my first time running a prerelease build of sandboxie.

![image](https://github.com/sandboxie-plus/Sandboxie/assets/18254607/827c5463-1d73-499a-a21f-fcb8dbf52b1e)

**Comment by DavidXanatos** [comment]
Yes you do, the driver in CI build is not signed 
so you eider need to enable test signin on your systeman reboot, 
or what usualy is enough use the signed driver from the last proper release.
As long as the driver does not change that works just as well

**Comment by isaak654** [comment]
I confirm that the fix is effective if you create a backup of the vault folder, delete the sandbox content, and restore the vault folder later.

**Comment by isaak654** [comment]
I confirm that the fix is effective if you create a copy of the vault folder, delete the sandbox content, and restore the vault folder later.

**Comment by profucius** [comment]
I tried the testsigning command and rebooted. I was able to run the portable sandboxie, however my sandboxies could not run any apps. Nothing would open and a WerFault.exe ran for a few seconds before everything closed inside SandMan. Even tried running the obsidian installer in a fresh sandbox, same issue. It's plausible this is related to my setup somehow, however when I reverted everything back to official 1.12.7 exes ran fine again. I can try again on my other fresh machine later when I get time.

@isaak654 Are you saying you followed these steps out of precaution, or that it was mandatory to do the copy/restore to get it working?

**Comment by isaak654** [comment]
I had to follow the above steps because Obsidian stopped opening in the sandbox after applying the CI build.

**Comment by isaak654** [comment]
I had to follow the above steps because Obsidian stopped opening in the sandbox after applying the CI build over the official 1.12.7.

**Comment by offhub** [comment]
I haven't tried it with programs other than browsers, but browsers with symlinked profile folders no longer open in the sandbox after this fix. (1.12.8)

Setting `UseNewSymlinkResolver=n` does not solve the problem.

```
@echo off
:: Run as Admin

if exist "%TMP%\Microsoft\Edge\User Data_3481\" (
	rd /s /q "%TMP%\Microsoft\"
	rd "%LocalAppData%\Microsoft\Edge\User Data\"
	ren "%LocalAppData%\Microsoft\Edge\User Data_3481_BAK" "User Data"
) else (
	xcopy /e "%LocalAppData%\Microsoft\Edge\User Data\" "%TMP%\Microsoft\Edge\User Data_3481\"
	ren "%LocalAppData%\Microsoft\Edge\User Data" "User Data_3481_BAK"
	mklink /D "%LocalAppData%\Microsoft\Edge\User Data" "%TMP%\Microsoft\Edge\User Data_3481"
)
```

https://github.com/sandboxie-plus/Sandboxie/assets/6871698/9f8b7f15-7389-4a70-8035-da7a0952d1ad

**Comment by DavidXanatos** [comment]
@offhub 
pleae remove UseNewSymlinkResolver=n 
the goal is for all to work with UseNewSymlinkResolver=y which is the default
UseNewSymlinkResolver=n  is the old sbie behaviore which is bugged

**Comment by offhub** [comment]
I've tested without `UseNewSymlinkResolver=n`.

**Comment by isaak654** [comment]
I wonder if it wouldn't be possible to write unit tests to be run as part of the workflow steps, otherwise we'll keep finding these regressions too late.

**Comment by isaak654** [comment]
I wonder if it wouldn't be possible to write unit tests to be run as part of the workflow steps, otherwise we'll keep chasing these regressions all the time.

**Comment by isaak654** [comment]
I wonder if it wouldn't be possible to write unit tests to be run as part of the workflow steps, otherwise we risk chasing these regressions all the time.

**Comment by isaak654** [comment]
I wonder if it wouldn't be possible to write unit tests to be run as part of the workflow steps, otherwise the risk is chasing these regressions all the time.

**Comment by JaGitU** [comment]
Seems SymlinkResolver is broken in the newest version (5.67.8)
The following batch worked fine before. Launch it under default sandbox and try to open SymlinkDir created in the opened window, now it crashes.

mklink /j "%~dp0SymLinkDir" "%~dp0"
explorer "%~dp0SymLinkDir"

**Comment by DavidXanatos** [comment]
please try this hitfix build: https://github.com/sandboxie-plus/Sandboxie/releases/download/v1.12.8/Sandboxie-Plus-x64-v1.12.8b.exe

**Comment by offhub** [comment]
Looks good to me.

**Comment by JaGitU** [comment]
> please try this hitfix build: https://github.com/sandboxie-plus/Sandboxie/releases/download/v1.12.8/Sandboxie-Plus-x64-v1.12.8b.exe

Now it’s good. Just using the opportunity would like to pay you attention on old unresolved issue #3178 that still exists (sorry for off-topic:)), hope it will be fixed in the future builds somehow.

**Comment by profucius** [comment]
> please try this hitfix build: https://github.com/sandboxie-plus/Sandboxie/releases/download/v1.12.8/Sandboxie-Plus-x64-v1.12.8b.exe

I'll give this a try tomorrow; I had some trouble with my other machine disabling driver signing but will give it another go.

**Comment by DavidXanatos** [comment]
1.12.8b has a signed driver only the installer and sbiedll.dll are unsigned

**Comment by profucius** [comment]
I just retested using the OpenFilePath instructions above. I was able to get everything working as expected on 1.12.8b. With a caveat

> OpenFilePath _instead_ of Symlink

This does not work. I put the directories in and deleted the symbolic links, and Obsidian app was completely empty. I added the symbolic links in again, and now when I edit the files within Obsidian, it writes directly to the original files and does not create duplicates/forks of them.

Going forward I will use OpenFilePath and Symbolic links in tandem to fulfill these needs. If it remains stable like this, then I will consider it resolved from my end.
